### DISCUSSION TITLE
Avoid re-compilation with every new image

### DISCUSSION THREAD

**Discussion by hoerenj** [halide/Halide#5888] [Q&A] [answered]
Beginner question: I've setup a pipeline that processes incoming images. Obviously, it is recompiled with every new frame. Is this necessary? How can I avoid this without moving to AOT? The pipeline should compile just once, and then process the incoming frames as fast as possible. I have worked through most of the tutorials, it seems most of them use a single input image only. Is there any simple example?

**Comment by zvookin** [comment, marked as answer]
The code is likely redefining the realized ```Func``` or ```Pipeline``` for each frame. If the ```Func``` is defined, or built, once and reused, it will be compiled once the first time it is called via ```realize```.

**Reply by hoerenj** [reply]
In my case, `Func` depends on `input`, a `Halide::Buffer` input buffer. To reuse `Func` (and to avoid re-compilation), I am currently copying any new incoming data to `input`, which is costly. Can I reuse `Func` w/o the copying, i.e. realize `Func` with a different input buffer? (In tutorial 10, it is shown for AOT mode, but I can't figure it out for JIT mode)

**Reply by zvookin** [reply]
You likely want an ```ImageParam```. The ```set``` method of the ```ImageParam``` assigns a buffer to the conceptual parameter. This happens without copying the data.

**Reply by hoerenj** [reply]
That's what I was looking for. Thanks!

**Reply by HelloHalide** [reply]
hi, can you paste your solution about this question?
